### PR TITLE
Parsing quantized nodes does not rely on names

### PR DESCRIPTION
**Pull request by zihaomu**
Related issues #22509 and #22442.
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 1]
```diff
@@ -754,49 +754,35 @@ void ONNXImporter::parseOperatorSet()
...
-        layerParams.set("scales", DictValue::arrayReal(scale.ptr<float>(), 1));
-        layerParams.set("zeropoints", DictValue::arrayInt(zeropoint.ptr<int8_t>(), 1));
+        if (layerType == "Quantize"
+            || (len > 3 && layerType.substr(len - 4) == "Int8"))
+        {
+            layerParams.set("depth", CV_8S);
+            return;
+        }
     }
+    layerParams.set("depth", CV_32F);
```
Comment (zihaomu) [inline modules/dnn/src/onnx/onnx_importer.cpp:785]: This code wants to get the depth of the current layer based on the node's input. Do we have a better way to get `depth` directly from the layer?

**Comment by asmorkalov** [comment]
@zihaomu I propose to touch some of quantized models (per-layer test) to cover new case with test too.

**Comment by zihaomu** [comment]
> @zihaomu I propose to touch some of quantized models (per-layer test) to cover new case with test too.

Hi @asmorkalov, thanks for your code reviewing. This patch wants to stop rely on the node name when we parse the quantized node. And it doesn't introduce new features. In my opinion, the existing test cases are enough.

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 2]
```diff
@@ -3113,17 +3099,19 @@ void ONNXImporter::parseCustomLayer(LayerParams& layerParams, const opencv_onnx:
 
 void ONNXImporter::parseQuantDequant(LayerParams& layerParams, const opencv_onnx::NodeProto& node_proto)
 {
-    CV_Assert(node_proto.input_size() == 3);
+    CV_Assert(node_proto.input_size() == 2 || node_proto.input_size() == 3);
     layerParams.type = (node_proto.op_type() == "QuantizeLinear") ? "Quantize" : "Dequantize";
+    Mat scale = getBlob(node_proto, 1);
+    int8_t zeropoint = node_proto.input_size() == 3 ? getBlob(node_proto, 2).at<int8_t>(0) : 0;
```
Comment (rogday) [inline modules/dnn/src/onnx/onnx_importer.cpp:3105]: Per spec, zeropoint should be of the same shape as scale, which can be either 0-d(per Tensor) or 1-d(per axis), but we don't support the latter: `CV_CheckEQ(scale.total(), 1, "")` should suffice.

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 3]
```diff
@@ -3293,8 +3289,10 @@ void ONNXImporter::parseQGemm(LayerParams& layerParams, const opencv_onnx::NodeP
             CV_Error(Error::StsUnsupportedFormat, "The zero-point non-zero case of W is not supported!");
     }
 
-    Mat out_sc, bias;
-    out_sc = getBlob(node_proto, 7);
+    Mat out_sc = getBlob(node_proto, 7);
+    Mat out_zp = getBlob(node_proto, 8);
```
Comment (rogday) [inline modules/dnn/src/onnx/onnx_importer.cpp:3293]: These are optional, so should be handled just like `bias` below.

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 4]
```diff
@@ -3183,6 +3171,8 @@ void ONNXImporter::parseQConv(LayerParams& layerParams, const opencv_onnx::NodeP
     Mat wt_sc = (w_scale.total() == outCn) ? w_scale : Mat(1, outCn, CV_32F, Scalar(w_scale.at<float>(0)));
 
     Mat out_sc = getBlob(node_proto, 6);
+    Mat out_zp = getBlob(node_proto, 7);
+    CV_Assert(out_sc.total() == 1 && out_zp.total() == 1);
```
Comment (rogday) [inline modules/dnn/src/onnx/onnx_importer.cpp:3175]: These are scalars per spec already, but I think we should add this assertion everywhere

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 5]
```diff
@@ -3227,6 +3220,7 @@ void ONNXImporter::parseQMatMul(LayerParams& layerParams, const opencv_onnx::Nod
     bool per_channel = w_scale.total() == outCn ? true : false;
     Mat wt_sc = (w_scale.total() == outCn) ? w_scale : Mat(1, outCn, CV_32F, Scalar(w_scale.at<float>(0)));
     Mat out_sc = getBlob(node_proto, 6);
+    Mat out_zp = getBlob(node_proto, 7);
```
Comment (rogday) [inline modules/dnn/src/onnx/onnx_importer.cpp:3223]: assertion missing: scalars

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 1 above)
Comment (rogday) [reply modules/dnn/src/onnx/onnx_importer.cpp:785]: Welp, layers have constrains on types in the spec, but we don't have the information in the graph which types this layer supports.

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 6]
```diff
@@ -754,49 +754,35 @@ void ONNXImporter::parseOperatorSet()
...
+            hasVariableInps = true;
+            break;
         }
+    }
 
-        if (scale.total() != 1 || zeropoint.total() != 1)
-            CV_Error(Error::StsNotImplemented, "Per-channel scales/zeropoints are not supported");
+    if (hasVariableInps)
+    {
+        LayerInfo layerInfo = layer_id.find(node_proto.input(i))->second;
```
Comment (rogday) [inline modules/dnn/src/onnx/onnx_importer.cpp:773]: So, you look for any real input layer and check if we created internal layer `*Int8`? I mean, someone could rename the layer or add a new one that doesn't follow that convention. Maybe we're better off with `layerParams.set("depth", CV_8S)` in each layer, or we could create a set of all integer layers. I think this should be discussed further with the rest of the team.

**Review by rogday — CHANGES_REQUESTED** [review]
Thank you for looking into this!

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 6 above)
Comment (zihaomu) [reply modules/dnn/src/onnx/onnx_importer.cpp:773]: Thanks for your code review. OpenCV DNN needs to allocate output blob according to depth type (for example the depth of `CV_8S` will be allocated as int8 Mat.)

> someone could rename the layer or add a new one that doesn't follow that convention.

It is possible. I agree with `create a set of all integer layers`.

Comment (rogday) [reply modules/dnn/src/onnx/onnx_importer.cpp:773]: I'm not sure relying on internal names is a good idea. Let's discuss this on Friday, maybe someone will come up with a better solution.

Comment (zihaomu) [reply modules/dnn/src/onnx/onnx_importer.cpp]: Hi @rogday, please take a look. I have refactored the code.

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 7]
```diff
@@ -754,49 +762,72 @@ void ONNXImporter::parseOperatorSet()
...
         {
-            std::string inpName = node_proto.input(0);
-            inpName = inpName.substr(0, inpName.length() - 9);
-            scale = getBlob(inpName + "scale");
-            zeropoint = getBlob(inpName + "zero_point");
+            LayerInfo layerInfo = layer_id.find(node_proto.input(i))->second;
+            Ptr<Layer> preLayer = dstNet.getLayer(layerInfo.layerId);
 
-            for (int i = 0; i < node_proto.output_size(); i++)
+            if (layerInfo.depth == CV_8S && hasInt8Input(preLayer->type))
```
Comment (rogday) [inline modules/dnn/src/onnx/onnx_importer.cpp:823]: So if we insert unsupported at the moment layer(Gather, for example) after integer node, it will set the depth to CV_8S and corrupt the memory at runtime? (reading int8 as float in this case).

Comment (rogday) [reply modules/dnn/src/onnx/onnx_importer.cpp:823]: What if we have a set of ONNX layers instead? e.g. in this set we would put `QuantizeLinear`, `DequantizeLinear`, `QLinearConv`, `Transpose`, `Reshape`, `MaxPool`, etc.
And check that if we have integer inputs, current layer should be in this set, otherwise we throw an error, what do you think?

**Review by rogday — COMMENTED** [review]
Thank you!

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 7 above)
Comment (zihaomu) [reply modules/dnn/src/onnx/onnx_importer.cpp:823]: > So if we insert unsupported at the moment layer(Gather, for example) after integer node, it will set the depth to CV_8S and corrupt the memory at runtime? (reading int8 as float in this case).

Yes, if we do this, it will corrupt. But this should not happen, Gather-like layer should output the same format as the input. So, if a Gather layer gets int8 input, it should be `GatherInt8`.

Comment (zihaomu) [reply modules/dnn/src/onnx/onnx_importer.cpp:823]: > What if we have a set of ONNX layers instead? e.g. in this set we would put QuantizeLinear, DequantizeLinear, QLinearConv, Transpose, Reshape, MaxPool, etc.
And check that if we have integer inputs, the current layer should be in this set, otherwise we throw an error, what do you think?

I also considered a list of ONNX operators. Functionally, there is no difference between the two lists in essence, one is the original version of ONNX and the other is the mapped version of OpenCV.
Considering the future maintainability, indeed using the original version of ONNX operator names is better.

**Comment by asmorkalov** [comment]
@zihaomu friendly reminder.

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 8]
```diff
@@ -3639,8 +3703,8 @@ void ONNXImporter::buildDispatchMap_ONNX_AI(int opset_version)
     dispatch["MaxPool"] = &ONNXImporter::parseMaxPool;
     dispatch["AveragePool"] = &ONNXImporter::parseAveragePool;
     dispatch["GlobalAveragePool"] = dispatch["GlobalMaxPool"] = &ONNXImporter::parseGlobalPool;
-    dispatch["ReduceMax"] = dispatch["ReduceMin"] = dispatch["ReduceMean"] = dispatch["ReduceSum"] = dispatch["ReduceMax"] =
-            dispatch["ReduceMin"] = dispatch["ReduceSumSquare"] = dispatch["ReduceProd"] = dispatch["ReduceL1"] =
+    dispatch["ReduceMax"] = dispatch["ReduceMin"] = dispatch["ReduceMean"] = dispatch["ReduceSum"] =
+            dispatch["ReduceSumSquare"] = dispatch["ReduceProd"] = dispatch["ReduceL1"] =
```
Comment (rogday) [inline modules/dnn/src/onnx/onnx_importer.cpp:3707]: What is the reason behind this change?

**Review by rogday — COMMENTED** [review]
Looks good, one question.

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 8 above)
Comment (zihaomu) [reply modules/dnn/src/onnx/onnx_importer.cpp:3707]: Hi @rogday, I found there this code is duplicated. There are two "ReduceMax".

Comment (rogday) [reply modules/dnn/src/onnx/onnx_importer.cpp:3707]: Oh, right, my bad.

**Review by rogday — APPROVED** [review]
LGTM! :+1:

**Comment by asmorkalov** [comment]
Discussed testing offline. Decided to merge without test.
